### PR TITLE
[DirectX][NFC] Change specification of overload types and attribute in DXIL.td

### PR DESCRIPTION
**Pull request by bharadwajy**
- Specify overload types of DXIL Operation as list of types instead of a string.
 - Add supported DXIL type record definitions to `DXIL.td` leveraging `LLVMType` to avoid duplicate definitions.
 - Spell out DXIL Operation Attribute specification string.
 - Make corresponding changes to process the records in DXILEmitter.cpp

### CODE REVIEW

**Comment by bharadwajy** [comment]
@llvm-beanz @bogner @dmpots @python3kgae

File: llvm/utils/TableGen/DXILEmitter.cpp [hunk 1]
```diff
@@ -268,36 +296,45 @@ static void emitDXILIntrinsicMap(std::vector<DXILOperationDesc> &Ops,
 
 static std::string emitDXILOperationFnAttr(StringRef FnAttr) {
   return StringSwitch<std::string>(FnAttr)
-      .Case("rn", "Attribute::ReadNone")
-      .Case("ro", "Attribute::ReadOnly")
+      .Case("NoReadMemory", "Attribute::ReadNone")
+      .Case("ReadMemory", "Attribute::ReadOnly")
```
Comment (python3kgae) [inline llvm/utils/TableGen/DXILEmitter.cpp:300]: Why not just ReadNone and ReadOnly?

Comment (bharadwajy) [reply llvm/utils/TableGen/DXILEmitter.cpp:300]: > Why not just ReadNone and ReadOnly?

The thought was to emphasize *memory* read, but I am open to changing if the terms chosen don't sound palatable.

Comment (python3kgae) [reply llvm/utils/TableGen/DXILEmitter.cpp:300]: ReadMemory lost the only part of ReadOnly, but ReadOnlyMemory feels bad too :(
Maybe OnlyReadMemory?

Comment (bharadwajy) [reply llvm/utils/TableGen/DXILEmitter.cpp:300]: > ReadMemory lost the only part of ReadOnly, but ReadOnlyMemory feels bad too :( Maybe OnlyReadMemory?

How about `MemoryReadNone` and `MemoryReadOnly`?

Comment (dmpots) [reply llvm/utils/TableGen/DXILEmitter.cpp:300]: Don't these map directly to the llvm function attributes? If so I think we should match those names (`readonly` and `readnone`).

Comment (bharadwajy) [reply llvm/utils/TableGen/DXILEmitter.cpp:300]: > Don't these map directly to the llvm function attributes? 

Yes. 

> If so I think we should match those names (`readonly` and `readnone`). Using the decided naming convention (e.g. CamlCase).

As the `DXIL.td` is expected to be the eventual single source of reference for DXIL operations and affords declarative specification, the thought was to be a bit more descriptive than C++ sources, when possible.

I can change them as suggested.

Comment (bharadwajy) [reply llvm/utils/TableGen/DXILEmitter.cpp]: I changed the type of `Attribute` to `EnumAttr` type to get the benefit of appropriately constrained values that can be specified and validating at compile-time as opposed to using a `string`.

**Review by python3kgae — APPROVED** [review]

File: llvm/lib/Target/DirectX/DXIL.td [hunk 2]
```diff
@@ -34,12 +35,42 @@ def BinaryUintCategory : DXILOpCategory<"Binary uint">;
 def UnaryFloatCategory : DXILOpCategory<"Unary float">;
 def ComputeIDCategory : DXILOpCategory<"Compute/Mesh/Amplification shader">;
 
+// ValueTypes specific to DXIL
+// Define Overload value type as an entity with no size and an arbitrary value
+// of 1024 - assuming that all currently defined values are less than 1024
```
Comment (dmpots) [inline llvm/lib/Target/DirectX/DXIL.td:40]: What happens if LLVM changes to have a value >= 1024? Is it going to break in a easily recognizable way that is obvious to fix? How close are we to that limit?

File: llvm/utils/TableGen/DXILEmitter.cpp [hunk 3]
```diff
@@ -267,38 +296,51 @@ static void emitDXILIntrinsicMap(std::vector<DXILOperationDesc> &Ops,
...
-static std::string emitDXILOperationFnAttr(StringRef FnAttr) {
-  return StringSwitch<std::string>(FnAttr)
-      .Case("rn", "Attribute::ReadNone")
-      .Case("ro", "Attribute::ReadOnly")
+// Convert operation attribute string to Attribute enum
+//
+// @param Attr string reference
+// @return std::string Attribute enum string
+static std::string emitDXILOperationAttr(StringRef Attr) {
+  return StringSwitch<std::string>(Attr)
```
Comment (dmpots) [inline llvm/utils/TableGen/DXILEmitter.cpp:304]: I see this is just following the existing code, but should we add some error handling for unexpected attributes here?

**Review by dmpots — APPROVED** [review]

File: llvm/lib/Target/DirectX/DXIL.td (hunk 2 above)
Comment (bharadwajy) [reply llvm/lib/Target/DirectX/DXIL.td]: > What happens if LLVM changes to have a value >= 1024? Is it going to break in a easily recognizable way that is obvious to fix? How close are we to that limit?

I have not investigated this. I am looking at a different way to define overload types without the reliance above as I share the same concerns. The current changes do not require these definitions. So, I have pushed a change that deletes these additions from this PR. Thanks!

**Pull request merged**